cms/src/main/java/uk/nhs/digital/externalstorage/modules/S3ConnectorServiceRegistrationModuleParams.java

```diff
@@ -1,74 +1,91 @@
-        "externalstorage.aws.s3.endpoint", "s3Endpoint", stringValue -> stringValue, ""
-    );
-
-    // DOWNLOAD PARAMS
-
-    static final S3ConnectorServiceRegistrationModuleParams<Integer> DOWNLOAD_MAX_CONC_COUNT = init(
-        "externalstorage.download.maxConcurrentDownloads", "maxConcurrentDownloadsCount", Integer::valueOf, 0
-    );
-    static final S3ConnectorServiceRegistrationModuleParams<Long> DOWNLOAD_SHUTDOWN_TIMEOUT_IN_SECS = init(
-        "externalstorage.download.shutdown.timeoutInSecs", "downloadShutdownTimeoutInSecs", Long::valueOf, 0L
-    );
-
-    // UPLOAD PARAMS
-
-    static final S3ConnectorServiceRegistrationModuleParams<Integer> UPLOAD_MAX_CONC_COUNT = init(
-        "externalstorage.upload.maxConcurrentUploads", "maxConcurrentUploadsCount", Integer::valueOf, 0
-    );
-    static final S3ConnectorServiceRegistrationModuleParams<Long> UPLOAD_SHUTDOWN_TIMEOUT_IN_SECS = init(
-        "externalstorage.upload.shutdown.timeoutInSecs", "uploadShutdownTimeoutInSecs", Long::valueOf, 0L
-    );
-
-    private final String systemPropertyKey;
-    private final String jcrModulePropertyKey;
-    private final Function<String, T> fromStringConverter;
-    private final T defaultValue;
-
-    private S3ConnectorServiceRegistrationModuleParams(final String systemPropertyKey,
-                                                       final String jcrModulePropertyKey,
-                                                       final Function<String, T> fromStringConverter,
-                                                       final T defaultValue
-    ) {
-        this.systemPropertyKey = systemPropertyKey;
-        this.jcrModulePropertyKey = jcrModulePropertyKey;
-        this.fromStringConverter = fromStringConverter;
-        this.defaultValue = defaultValue;
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+
+public class S3ConnectorServiceRegistrationModuleParams {
+    private String s3Bucket;
+    private String s3Region;
+    private String s3Endpoint;
+
+    private long downloadsMaxConcurrentCount;
+    private long downloadsShutdownTimeoutInSecs;
+
+    private long uploadsMaxConcurrentCount;
+    private long uploadsShutdownTimeoutInSecs;
+
+    // SYSTEM PROPERTY KEYS
+    public static final String KEY_AWS_BUCKET =  "externalstorage.aws.bucket";
+    public static final String KEY_AWS_REGION =  "externalstorage.aws.region";
+    public static final String KEY_AWS_S3_ENDPOINT =  "externalstorage.aws.s3.endpoint";
+
+    // JCR PROPERTIES KEYS
+    public static final String KEY_DOWNLOAD_MAX_CONC_COUNT =  "maxConcurrentDownloadsCount";
+    public static final String KEY_DOWNLOAD_SHUTDOWN_TIMEOUT_IN_SECS =  "downloadShutdownTimeoutInSecs";
+    public static final String KEY_UPLOAD_MAX_CONC_COUNT =  "maxConcurrentUploadsCount";
+    public static final String KEY_UPLOAD_SHUTDOWN_TIMEOUT_IN_SECS =  "uploadShutdownTimeoutInSecs";
+
+    S3ConnectorServiceRegistrationModuleParams(final Node moduleConfigNode) {
```

Comment:
I know I initially suggested placing the logic extracting the values in the constructor, but thinking about it it would probably better to:
* move this logic into a static public factory method (`extractParameters`? `newInstance`?), taking the `Node` as its parameter),
* make the constructor take the full list of already extracted values,
* make the constructor private.

This leaves the constructor with no other logic than just populating the fields which is typically a good practice.

Reply:
Now updated.